index.ts

```diff
@@ -14,18 +14,18 @@
  * to cleanly shut down before the process exits.
  */
 
-// TODO(rictic): convert this library to promises.
-
 /**
  * The type of a cleankill interrupt handler.
  */
-export type Handler = (doneHandling: () => void) => void;
+export type Handler = () => Promise<void>;
 
-let interruptHandlers: Handler[] = [];
+const interruptHandlers: Handler[] = [];
 
 /**
- * Register a handler to occur on SIGINT. All handlers are passed a callback,
- * and the process will be terminated once all handlers complete.
+ * Register a handler to occur on SIGINT. The handler must return a promise if
```

Comment:
In the naive method in this PR those two examples would be:

```js
// synchronous handler
cleankill.onInterrupt(() => this.cleanupSync());
```

```js
// asynchronous handler
cleankill.onInterrupt(async () => {
  await this.cleanupAsync();
});
```

This seems better because there's no special object with a `waitFor` method or anything, and error propagation along every codepath is very straightforward. I admit that I may be missing some of the virtues of your proposal though.